source/extensions/filters/http/transformation/inja_transformer.cc

```diff
@@ -244,6 +259,41 @@ json TransformerInstance::cluster_metadata_callback(
   return "";
 }
 
+json TransformerInstance::base64_encode_callback(const inja::Arguments &args) const {
+  const std::string &input = args.at(0)->get_ref<const std::string &>();
+  return Base64::encode(input.c_str(), input.length());
+}
+
+json TransformerInstance::base64_decode_callback(const inja::Arguments &args) const {
+  const std::string &input = args.at(0)->get_ref<const std::string &>();
+  return Base64::decode(input);
+}
+
+json TransformerInstance::substring_callback(const inja::Arguments &args) const {
+  const std::string &input = args.at(0)->get_ref<const std::string &>();
+  const int64_t start = args.at(1)->get_ref<const int64_t &>();
```

Comment:
can we add a test for what happens here if a user provides a non-integer arg?

Reply:
good call, will do

Reply:
Added this here: https://github.com/solo-io/envoy-gloo/pull/184/commits/7e36cc2559e7a994f249aecd1f8276d67375cd3c

A couple of notes
 - The pattern in the custom Inja callbacks we define here seems to be that we return an empty string if there is invalid input/we can't find what the input points to. See for example: https://github.com/solo-io/envoy-gloo/blob/8ddf225f48d8db0fc7f86aa9c62411797967568e/source/extensions/filters/http/transformation/inja_transformer.cc#L169-L177
   - I think this makes sense intuitively, because the callbacks are supposed to be composable, so returning an error string or throwing an exception is probably not desirable behavior
 - The substring callback now returns an empty string if the second argument cannot be parsed as an integer
 - Probably more open to interpretation, the substring callback now also returns an empty string if there is a third argument and it cannot be parsed as an integer
   - How do you feel about this? Do you think we should just act as though there is no third argument in this case?

Reply:
Also, the envoy base64 utilities we call into will return an empty string if we attempt to decode invalid base64. I'm going add a test against this as well
https://github.com/envoyproxy/envoy/blob/main/source/common/common/base64.cc#L146-L151

Reply:
Added this here: f105915196798ca381358177f2c9e34070cf2919